### PR TITLE
Canvas 26 - Handle crosslisted courses.

### PR DESCRIPTION
**Pull request by kilometerskang**

### CODE REVIEW

File: src/server/services/registrar.js [hunk 1]
```diff
@@ -203,8 +204,43 @@ async function getShortname(offeredTermCode, classSectionID) {
...
+ * @param {string} classSectionID   ClassID aka SRS.
+ * @returns {Array}   Returns a list of crosslisted courses.
+ */
+async function getCrosslists(offeredTermCode, classSectionID) {
+  const crosslistXml = await axios(
+    `https://webservices.registrar.ucla.edu/SRDB/SRDBWeb.asmx/getConSched?user=${process.env.REG_DB_USER}&pass=${process.env.REG_DB_PASS}&term=${offeredTermCode}&SRS=${classSectionID};`
+  );
+
+  const crosslists = [];
+  xml2js.parseString(crosslistXml.data, (err, result) => {
```
Comment (finnzink) [inline src/server/services/registrar.js:220]: I believe axios handles xml parsing automatically. Might make code slightly cleaner

Comment (kilometerskang) [reply src/server/services/registrar.js:220]: I might be wrong but I couldn't find any docs on that...

File: src/server/services/registrar.js [hunk 2]
```diff
@@ -203,8 +204,43 @@ async function getShortname(offeredTermCode, classSectionID) {
...
+        crosslist.srs[0]
+      );
+      if (crosslistShortname) {
+        crosslists.push({
+          term: crosslist.term[0],
+          srs: crosslist.srs[0],
+          shortname: crosslistShortname,
+        });
+      } else {
+        registrarDebug('getShortname returned null');
```
Comment (rlorenzo) [inline src/server/services/registrar.js:233]: @kilometerskang For this debugging info to be useful we need to know the term/srs in which it returned null

File: src/server/services/registrar.js [hunk 3]
```diff
@@ -203,8 +204,43 @@ async function getShortname(offeredTermCode, classSectionID) {
...
+
+  const crosslists = [];
+  xml2js.parseString(crosslistXml.data, (err, result) => {
+    result.ArrayOfGetConSchedData.getConSchedData.forEach(async crosslist => {
+      const crosslistShortname = await registrar.getShortname(
+        crosslist.term[0],
+        crosslist.srs[0]
+      );
+      if (crosslistShortname) {
+        crosslists.push({
```
Comment (rlorenzo) [inline src/server/services/registrar.js:227]: @kilometerskang Also add registrarDebug() for how many and what crosslists were found

**Comment by rlorenzo** [comment]
@kilometerskang I see this warning in the jest test: https://github.com/ucla/lti-library-resources/pull/14/checks?check_run_id=1004401968#step:6:10

Was that introduced in this pull request? Can it be fixed regardless?

File: src/client/components/Nav/test.js [hunk 4]
```diff
@@ -5,6 +5,9 @@ import Nav from '.';
 it('Tabs render correctly for clusters', () => {
```
Comment (rlorenzo) [inline src/client/components/Nav/test.js:5]: @kilometerskang The filename is really generic. Can it be renamed so it reflects what this test is actually doing? Like named after the tab it is testing?

Comment (kilometerskang) [reply src/client/components/Nav/test.js:5]: All my test files are named this way.. I figured the folder it is in is explanatory enough. This file is also just testing the entire Nav component in general, no tab specifically

**Comment by kilometerskang** [comment]
I

**Comment by kilometerskang** [comment]
> @kilometerskang I see this warning in the jest test: https://github.com/ucla/lti-library-resources/pull/14/checks?check_run_id=1004401968#step:6:10
> 
> Was that introduced in this pull request? Can it be fixed regardless?

The Cluster ticket takes care of this warning. Should I just fix it here too? it might be easier to leave it because it would cause merge conflicts to have 2 different fixes for it

**Comment by rlorenzo** [comment]
> > @kilometerskang I see this warning in the jest test: https://github.com/ucla/lti-library-resources/pull/14/checks?check_run_id=1004401968#step:6:10
> > Was that introduced in this pull request? Can it be fixed regardless?
> 
> The Cluster ticket takes care of this warning. Should I just fix it here too? it might be easier to leave it because it would cause merge conflicts to have 2 different fixes for it

Okay, just fix it in the Cluster ticket

File: src/server/services/registrar.js [hunk 5]
```diff
@@ -203,8 +204,46 @@ async function getShortname(offeredTermCode, classSectionID) {
...
+ */
+async function getCrosslists(offeredTermCode, classSectionID) {
+  const crosslistXml = await axios(
+    `https://webservices.registrar.ucla.edu/SRDB/SRDBWeb.asmx/getConSched?user=${process.env.REG_DB_USER}&pass=${process.env.REG_DB_PASS}&term=${offeredTermCode}&SRS=${classSectionID};`
+  );
+
+  const crosslists = [];
+  xml2js.parseString(crosslistXml.data, (err, result) => {
+    result.ArrayOfGetConSchedData.getConSchedData.forEach(async crosslist => {
+      const crosslistShortname = await registrar.getShortname(
```
Comment (rlorenzo) [inline src/server/services/registrar.js:222]: @kilometerskang You need to check if there are no crosslists. This code is calling getShortname() even if there are no crosslists.

File: src/server/services/registrar.js [hunk 6]
```diff
@@ -203,8 +204,46 @@ async function getShortname(offeredTermCode, classSectionID) {
...
+        registrarDebug(
+          `getShortname returned null for term ${crosslist.term[0]} and SRS ${crosslist.srs[0]}`
+        );
+      }
+    });
+  });
+  registrarDebug(`Crosslists found: ${crosslists}`);
+
+  return crosslists;
+}
```
Comment (rlorenzo) [inline src/server/services/registrar.js:242]: @kilometerskang I tested this function and it always returns an empty array even if there are crosslists.

For example if in src/server/jobs/checkRegistrarWS.js you put in this code:

```
(async () => {
  const result = await registrar.getCrosslists('20S', '250822200');

  console.log(result);
})();
```

It returns []. But it should return an array with term/srs plus shortnames 20S-BIOENGR287CM-1, 20S-COMSCI287CM-1, 20S-COMSCI187CM-1

**Pull request merged**